benchmarks/functional_autograd_benchmark/functional_autograd_benchmark.py

```diff
@@ -0,0 +1,144 @@
+import torch
+from torch.autograd import functional
+
+import time
+import argparse
+from collections import namedtuple, defaultdict
+
+import ppl_models
+import vision_models
+import audio_text_models
+
+from utils import get_str
+
+# Listing of the different tasks
+FAST_TASKS_NO_DOUBLE_BACK = [
+    "vjp",
+]
+
+FAST_TASKS = FAST_TASKS_NO_DOUBLE_BACK + [
+    "vhp",
+    "jvp",
+]
+
+ALL_TASKS = FAST_TASKS + [
+    "hvp",
+    "jacobian",
+    "hessian"
+]
+
+DOUBLE_BACKWARD_TASKS = ["jvp", "hvp", "vhp", "hessian"]
+
+# Model definition which contains:
+# - name: a string with the model name.
+# - getter: a function to get the mode. It takes as input the device on which the model
+#     will run. It should return the forward function and the parameters (Tensors) used as
+#     input for the forward function. Note that the forward must *not* have any side effect.
+# - tasks: the list of recommended tasks that can run in a reasonable amount of time with this model.
+# - unsupported: the list of tasks that this model cannot run.
+ModelDef = namedtuple("ModelDef", ["name", "getter", "tasks", "unsupported"])
+
+MODELS = [
+    ModelDef("resnet18", vision_models.get_resnet18, FAST_TASKS, []),
+    ModelDef("fcn_resnet", vision_models.get_fcn_resnet, FAST_TASKS, []),
+    ModelDef("detr", vision_models.get_detr, FAST_TASKS, []),
+    ModelDef("ppl_simple_reg", ppl_models.get_simple_regression, ALL_TASKS, []),
+    ModelDef("ppl_robust_reg", ppl_models.get_robust_regression, ALL_TASKS, []),
+    ModelDef("wav2letter", audio_text_models.get_wav2letter, FAST_TASKS, []),
+    ModelDef("deepspeech", audio_text_models.get_deepspeech, FAST_TASKS_NO_DOUBLE_BACK, DOUBLE_BACKWARD_TASKS),
+    ModelDef("transfo", audio_text_models.get_transformer, FAST_TASKS, []),
+    ModelDef("multiheadattn", audio_text_models.get_multiheadattn, FAST_TASKS, []),
+]
+
+def run_once(model, inp, task, extra=None):
+    func = getattr(functional, task)
+
+    # Cache the creation of the `v` argument
+    if extra is None:
+        if task in ["vjp"]:
+            out = model(*inp)
+            v = torch.rand_like(out)
+        elif task in ["jvp", "hvp", "vhp"]:
+            if isinstance(inp, tuple):
+                v = tuple(torch.rand_like(i) for i in inp)
+            else:
+                v = torch.rand_like(inp)
+        else:
+            v = None
+    else:
+        v = extra
+
+    if v is not None:
+        res = func(model, inp, v=v, strict=True)
+    else:
+        res = func(model, inp, strict=True)
+
+    if extra is None:
+        return v
+
+
+def run_model(model_getter, args, task):
+    if args.gpu == -1:
+        device = "cpu"
+
+        def noop(*args):
+            return args
+        do_sync = noop
+    else:
+        device = "cuda:{}".format(args.gpu)
+        do_sync = torch.cuda.synchronize
+
+    model, inp = model_getter(device)
+
+    # Warmup
+    extra = run_once(model, inp, task)
```

Comment:
```
extra = run_once(model, inp, task)
...
run_once(model, inp, task, extra)
```
is kind of hard to read because the name `extra` doesn't make much sense (maybe it should be `cached_v`?) and also it's weird that a function called `run_once` creates `v`. Maybe it would be better to separate the creation of v from `run_once`, e.g.

```
v = get_v_for(model, input, task)
run_once(model, inp, task, v)
...
run_once(model, inp, task, v)
```